### ISSUE TITLE
log4js causing issues with client-side usage

### ISSUE THREAD

**Issue by adrianmcli** [NaturalNode/natural#320]
I'm trying to use natural on the client side and I'm getting errors on (Webpack) compilation like the following:

```
Error in ./~/log4js/lib/appenders/clustered.js
Module not found: 'cluster' in /Users/liadrian/Dev/naive-bayes/node_modules/log4js/lib/appenders

 @ ./~/log4js/lib/appenders/clustered.js 3:14-32

Error in ./~/log4js/lib/appenders/gelf.js
Module not found: 'dgram' in /Users/liadrian/Dev/naive-bayes/node_modules/log4js/lib/appenders

 @ ./~/log4js/lib/appenders/gelf.js 5:12-28

Error in ./~/log4js/lib/appenders/hipchat.js
Module not found: 'hipchat-notifier' in /Users/liadrian/Dev/naive-bayes/node_modules/log4js/lib/appenders

... and many more like the above
```

### The problem

This is caused by a dependency on log4js, which for some reason has a lot of require statements in its appenders that are not declared in its own package.json file.

An example:

```js
// node_modules/log4js/lib/appenders/clustered.js
var cluster = require('cluster');
```

But `package.json` from the log4js module does not include `cluster` as a dependency. As a result, when Webpack is trying to compile my app, an error is thrown for each of these delinquent requires (i.e. require statements for modules not listed in its `package.json` dependency list).

The bottom line is: **log4js is the problem.**

### The solution

After looking into this, I was able to find that log4js is only used within the Brill POS Tagger module. I recommend that we move away from log4js completely and replace those logging statements with normal `console.log` or `console.warn` statements instead.

I have already attempted to do this, and can confirm that this will indeed solve the problem.

I will submit a PR if the community agrees it is beneficial to the project.

**Comment by adrianmcli** [comment]
FYI: I have created a PR to fix this issue #321.

**Comment by Hugo-ter-Doest** [comment]
Fine with me. You loose the possibility to configure (level of) logging. But the possibility to use natural in the browser is more valuable I think.

**Comment by adrianmcli** [comment]
@Hugo-ter-Doest Yeah, I realize that. It's unfortunate, but it was only ever implemented within the Brill POS Tagger anyway, so the loss isn't that huge. Please consider merging the PR once you've had a look.

I'm going to work hard on trying to make sure this library can work well on the client side.

**Comment by Hugo-ter-Doest** [comment]
One of the committers has to merge the PR, I cannot. But I support it.

Regards,
Hugo

**Comment by owendall** [comment]
👍 I just tried to browserify and ran into the same issue...

**Comment by owendall** [comment]
@adrianmcli :+1:

**Comment by adrianmcli** [comment]
@owendall glad to know it wasn't just me! Thanks for merging, @kkoch986!

**Comment by kkoch986** [comment]
np thanks for the pr!

**Comment by owendall** [comment]
@adrianmcli @kkoch986 Can you share your Webpack command for creating the standalone?  I just tried browserify but not getting the global I need for the browser?

**Comment by owendall** [comment]
I used:

`browserify index.js --standalone natural > natural.js --ignore lapack`

**Comment by owendall** [comment]
minified with uglify i get 380kb

**Comment by adrianmcli** [comment]
@owendall I don't really need to build a standalone module for natural as Webpack is a module bundler that'll bring it all in for me when I build my app.

That being said, I did try to make a simple demo Naive Bayes classifier app (with React), and the javascript bundle came out to **397 KB**. That's of course with a production build, so it's been minified and a few other optimizations automatically done for me.

But that file size includes the React runtime, all the code relating to my app and Natural itself. I don't believe Webpack 1 does any kind of tree-shaking, so this is actually really quite impressive.

**Comment by owendall** [comment]
@adrianmcli thx for the info.  So far I have been testing NLP Compromise V7, but looking at what might be more useful (e.g., Wordnet) for a few use cases:

https://github.com/nlp-compromise/nlp_compromise/tree/v7

**Comment by MichaelPaulukonis** [comment]
Is there any general discussion over making Natural client-side friendly?

While I don't generally oppose such initiatives, since there exists a JS NLP project specifically targeting the browser (`nlp-compromise`), I'd much rather remove that from consideration from Natural.

I'd much rather have a robust server-side/desktop tool that has more features, and another less-featured project that is optimized for a browser.

**Comment by adrianmcli** [comment]
@MichaelPaulukonis As great as NLP-Compromise is, it doesn't have all the features that Natural has, such as the classifiers.

More importantly, the fact that Natural CAN be used client-side if we follow best practices from some of the more popular libraries mean that we shouldn't exclude that from our roadmap.

The other consideration for this issue specifically is that we were depending on a Log4JS which had missing dependencies in its package.json. By removing this dependency, we improved the codebase. I believe there are other logging libraries which don't have this issue with the missing dependencies. Smart logging isn't limited to the server-side, after all.

In conclusion, if your goal is to have a robust library that works well on the server-side, then you should attempt to make it at least somewhat usable on the browser-side. If there's something we are using that only works on the server side, then maybe we are actually making our library LESS robust.

**Comment by moffsugita** [comment]
did you resolve?

**Comment by Hugo-ter-Doest** [comment]
This was fixed in #321
